query_executor.go

```diff
@@ -201,16 +210,375 @@ func (q *queryExecutor) do(ctx context.Context, qry ExecutableQuery, hostIter Ne
 	}
 
 	if lastErr != nil {
-		return &Iter{err: lastErr}
+		return newErrIter(lastErr, qry.getQueryMetrics())
 	}
 
-	return &Iter{err: ErrNoConnections}
+	return newErrIter(ErrNoConnections, qry.getQueryMetrics())
 }
 
-func (q *queryExecutor) run(ctx context.Context, qry ExecutableQuery, hostIter NextHost, results chan<- *Iter) {
+func (q *queryExecutor) run(ctx context.Context, qry internalRequest, hostIter NextHost, results chan<- *Iter) {
 	select {
 	case results <- q.do(ctx, qry, hostIter):
 	case <-ctx.Done():
 	}
-	qry.releaseAfterExecution()
+}
+
+type queryOptions struct {
```

Comment:
Shall we group here all relevant options into sub-types, for example:
```
Paging
    pageSize
    initialPageState
    disableAutoPage
Monitoring
    observer
    trace
Parameters
    values
    binding
WriteTimestamp
    defaultTimestamp
    defaultTimestampValue
Consistency
    initialConsistency
    serialConsistency
FrameOptions
    customPayload
    disableSkipMetadata

stmt
prefetch
rt
spec
context
idempotent
keyspace
skipPrepare
routingKey
```
Grouping is not definitive, it is just an idea.

Reply:
On a second thought I'm not sure if I like this idea that much, I'd have to create a few new types when maybe I can just group them a bit more by arranging the way they are declared... Wdyt